### PR TITLE
Version 0.2 - asyncio support.

### LINKED ISSUE #29: ASGI support
as mentioned in https://github.com/tomchristie/apistar/issues/9

### PR DESCRIPTION
**Pull request by tomchristie**
Pretty serious re-working in progress here. For an overview you'll want to review the source code of the branch itself rather than attempting to review the diff.

Take a look in particular at `interfaces.py` and `app.py`. You'll see that the dependency injection has been completely reworked, and that overriding components on applications looks far neater.

The end result of this work is going to be the ability for API Star to support multiple types of application. asyncio is the obvious first target here, but I'd also like us to have Flask and Django variants. Having the ability to use their ORMs as we currently have is okay, but users still miss out on a vast amount of existing ecosystem. 

**Update: This branch now introduces asyncio support.**

More work on this to come, but looking rather nice so far.

- [x] MultiDict for headers and queryparams.
- [x] Schema -> Components.
- [x] Finesse `interface.py`/`types.py`.
- [x] Finesse view names.
- [x] CLI support.
- [x] Component pre-loading.
- [x] Support defaults on params.
- [x] `Include` support.
- [x] Use `Routes` in preference to `Router` in components.
- [x] Reintroduce remaining commands, and `apistar` main cli.
- [x] Clean up response generation.
- [x] Coverage -> 100%
- [x] Simplify type->schema mapping.
- [x] Compile changelist.
- [x] Update docs.
- [x] Add asyncio docs.
- [x] Add Command docs.
- [x] Improve formatting of argparse commands.
- [x] Test API docs against 0.2
- [x] Finish remaining asyncio coverage.
- [x] Add `run` command for asyncio.
- [x] Case-insensitivity on headers.
- [x] Review `typing.newtype()` vs `typing.integer(...)` (Sticking with `newtype` for now)
- [x] Reintroduce ORM components.
- [x] Review statics/sendfile. Consider static docs generation.
- [x] `pip install apistar[asyncio]`, and update `apistar new --layout` to `apistar new --framework`
- [x] Reintroduce `test` command, at least for now.
- [x] Inline typesystem declarations.
- [x] Review settings.
- [ ] Finesse exception handling. 500 vs raise.
- [ ] CORS handling in finalize_response.
- [ ] Add backend testing docs for DB setup/teardown.

---
 Later:

* Remove `coreschema`, move form rendering into API Star. Add `render_form`, `render_link`.
* ~~Asyncio support.~~
* Flask support.
* Django support.
* ~~Reintroduce ORM components.~~
* `Schema` as a well-defined interface.
* Consider moving away from global settings.
* Make standalone usage of `app.injector` easier, eg. `static_files = app.injector.run(StaticFiles)`, update corresponding test cases.
* Better messages on injector configuration errors. (Use function and parameter name, not `Parameter`)
* Coerce primitive types to typesystem types for the purposes of stricter behavior (eg. inf) and validation error messages.
* Check for repeated list/dict in annotations, or list/dict on GET requests.
* Support explicit location overrides on parameters.
* Support form renderering list/dict type schemas fully, and schema output.
* Support type aliases as parameter annotations. typing.List, typing.NewType etc...
* Support defaults in params for HTTP.
* Error details on 404. Consider removing "messages" and always error by key.
* Finesse argparse support (use help/description generation for schemas too. support list/dict/enum)
* Support async context managers.
* Finesse Injector interface for sync/async, and raise error in sync case if coroutines are included.
* Add logging to components throughout.

---

Changelist: See README.

Closes #29

### CODE REVIEW

**Comment by ssteinerx** [comment]
@tomchristie Anything in here or the issue tracker that you'd specifically like help with?

**Comment by tomchristie** [comment]
@ssteinerx - Thanks! No not especially, the remaining work on here needs my attention specifically, because it's all design stuff. Where contributions will start to come into their own is once this API gets nailed down and we get to a point that third party components are an easy thing for folks to work on.

File: apistar/http.py [hunk 1]
```diff
@@ -1,263 +1,30 @@
...
-        if content_type and 'Content-Type' not in headers_dict:
-            headers_list += [('Content-Type', content_type)]
-
-        if status is None:
-            status = 200 if content_type else 204
-
-        self.data = data
+                 content: typing.Any,
+                 status: int=200,
+                 headers: typing.Dict[str, str]=None) -> None:
```
Comment (argaen) [inline apistar/http.py]: Can't `Headers` defined above be used here?

Comment (tomchristie) [inline apistar/http.py]: We'll likely want this annotation to actually be a `Union` of `dict/list of tuples/Headers`.
(ie. let the user use any of those, and then coerce them to Headers.
And yes, don't consider that annotation to be complete just yet, will get updated before 0.2 lands.

**Comment by argaen** [comment]
I'm trying to understand how `asyncio` would be integrated in v0.2. I have two guesses

1. Creating a new `Injector` that would be able to deal with coroutines in the `run` method?
2. The current `injector` is already able to run coroutines using the stepped functions (but I doubt that because I don't see any`step.send`).

3. I have no idea what I'm talking about -> It's also a possibility

Do you have your thoughts about this written in any blog post/docs or smth so I can read more about it?

**Comment by audiolion** [comment]
@argaen I would say we will see soon enough! You might want to make a post on discuss.apistar.org about it.

**Comment by tomchristie** [comment]
@argaen (1) is closest. The injector will need to have a `run_async` that:

1. Is an async function, rather than the regular function that we use for `run`
2. Checks if steps are coroutines, and calls them with await if they are.
3. (For bonus points) Support async context managers.

The server/application interface is different too of course - we can't use WSGI here. However the messaging interface and the server implementation work for that is already pretty much sorted out, here: https://github.com/encode/uvicorn

We'll also want:

* A different App implementation, so that we've got either eg. `from apistar.frameworks.wsgi import WSGIApp` or `from apistar.frameworks.asyncio import ASyncIOApp`.
* Have the TestClient be able to call into an asyncio app, from regular sync testing code.

**Comment by audiolion** [comment]
v0.2 is getting so close 🙌 ! Take my energy  ༼ つ ◕_◕ ༽つ

**Comment by tomchristie** [comment]
@audiolion Ha, yup! 😎 

Sound be out next week. We'll need to take a call on the ORM components. Right now they just don't exist in the fork. I suppose putting them back in as-is will probably be the right thing to do, tho I'd expect that in the future we may well push them into third party packages, and may also re-jig how settings work (I'd quite like to drop global settings and have the configuration pushed to the components).

**Comment by audiolion** [comment]
HOOOORAY ITS HERE 🎉 🎉

**Comment by tomchristie** [comment]
Few last things that I'd like to get sorted, but I'll defer them to a minor release bump later this week.
Let's get this out!

**Comment by tomchristie** [comment]
Alrighty, pushed to PyPI.

**Comment by audiolion** [comment]
To migrate from pre 0.2 is the best idea to just start over especially if its a small project?

**Comment by tomchristie** [comment]
@audiolion Probably review https://github.com/tomchristie/apistar#02-release and go from there. If I've missed anything let me know.

**Comment by tomchristie** [comment]
No doubt will discover some bits & pieces to clean up, and there's still some things I'm not sure I'm 100% about (eg. the import layouts), but the architecture is *far* cleaner now.

**Pull request merged**